execution/block-partitioner/src/v2/mod.rs

```diff
@@ -0,0 +1,750 @@
+            wsets.push(RwLock::new(HashSet::with_capacity(txn.write_hints().len())));
+            rsets.push(RwLock::new(HashSet::with_capacity(txn.read_hints().len())));
+        }
+        let timer_1 = MISC_TIMERS_SECONDS
+            .with_label_values(&["preprocess__main"])
+            .start_timer();
+        self.thread_pool.install(|| {
+            (0..num_txns)
+                .into_par_iter()
+                .for_each(|txn_idx: OriginalTxnIdx| {
+                    let txn = &txns[txn_idx];
+                    let sender = txn.sender();
+                    let sender_idx = *sender_idx_table
+                        .entry(sender)
+                        .or_insert_with(|| num_senders.fetch_add(1, Ordering::SeqCst));
+                    *senders[txn_idx].write().unwrap() = Some(sender_idx);
+                    let num_writes = txn.write_hints().len();
+                    for (i, storage_location) in txn
+                        .write_hints()
+                        .iter()
+                        .chain(txn.read_hints().iter())
+                        .enumerate()
+                    {
+                        let key = storage_location.state_key().clone();
+                        let key_idx = *key_idx_table
+                            .entry(key)
+                            .or_insert_with(|| num_keys.fetch_add(1, Ordering::SeqCst));
+                        let is_write = i < num_writes;
+                        if is_write {
+                            wsets[txn_idx].write().unwrap().insert(key_idx);
+                        } else {
+                            rsets[txn_idx].write().unwrap().insert(key_idx);
+                        }
+                        let tracker_ref = trackers
+                            .entry(key_idx)
+                            .or_insert_with(|| {
+                                let anchor_shard_id =
+                                    get_anchor_shard_id(storage_location, num_executor_shards);
+                                RwLock::new(ConflictingTxnTracker::new(
+                                    storage_location.clone(),
+                                    anchor_shard_id,
+                                ))
+                            });
+                        let mut tracker = tracker_ref
+                            .write()
+                            .unwrap();
+                        if is_write {
+                            tracker.add_write_candidate(txn_idx);
+                        } else {
+                            tracker.add_read_candidate(txn_idx);
+                        }
+                    }
+                });
+        });
+        let _duration_1 = timer_1.stop_and_record();
+        let _duration = timer.stop_and_record();
+
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["pre_partition_uniform"])
+            .start_timer();
+        let remaining_txn_idxs = uniform_partition(num_txns, num_executor_shards);
```

Comment:
This can be done as a follow up, but we need to create an abstraction for the partitioner here.. so that we can plug in other types of partitioner as a next evolution.